src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
+          malformed_flatpak_autoinstall_file_error (error,
+                                                    *autoinstall_flatpak_lines_iter,
+                                                    autoinstall_flatpak_file_path);
+          return NULL;
+        }
+
+      if (!parse_flatpak_ref_kind (flatpak_ref_kind, &flatpak_parsed_ref_kind))
+        {
+          eos_updater_remove_recursive (checkout_directory_path, NULL);
+          malformed_flatpak_autoinstall_file_error (error,
+                                                    *autoinstall_flatpak_lines_iter,
+                                                    autoinstall_flatpak_file_path);
+          return NULL;
+        }
+
+      g_ptr_array_add (autoinstall_flatpaks,
+                       flatpak_to_install_new_steal (g_steal_pointer (&flatpak_repo),
+                                                     g_steal_pointer (&flatpak_app_id),
+                                                     flatpak_parsed_ref_kind));
+    }
+
+  eos_updater_remove_recursive (checkout_directory_path, NULL);
+  return autoinstall_flatpaks;
+}
+
+static const gchar *
+get_gnome_software_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_GNOME_SOFTWARE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/gnome-software");
+}
+
+static GHashTable *
+read_flatpak_list_from_file_into_set (const gchar *path,
+                                      GError      **error)
+{
+  g_autofree gchar *contents = NULL;
+  g_auto(GStrv) strv = NULL;
+  g_autoptr(GHashTable) hash_set = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  GStrv strv_iter = NULL;
+
+  if (!g_file_get_contents (path, &contents, NULL, error))
+    {
+      /* Treat ENOENT as an empty-case */
+      if (g_error_matches (*error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_clear_error (error);
+          return g_steal_pointer (&hash_set);
+        }
+
+      return NULL;
+    }
+
+  strv = g_strsplit (contents, "\n", -1);
+
+  for (strv_iter = strv; *strv_iter; ++strv_iter)
+    g_hash_table_add (hash_set, g_strdup (*strv_iter));
+
+  return g_steal_pointer (&hash_set);
+}
+
+static GHashTable *
+determine_set_of_installed_applications (GError **error)
+{
+  const gchar *installed_flatpaks_list = g_getenv ("EOS_UPDATER_TEST_INSTALLED_FLATPAKS_LIST");
+
+  if (installed_flatpaks_list)
+    return read_flatpak_list_from_file_into_set (installed_flatpaks_list, error);
+
+  /* Not implemented for now, but read from the system flatpak installation */
+  return NULL;
+}
+
+static GPtrArray *
+resolve_flatpaks_to_deploy (GPtrArray *flatpaks_this_commit_wants,
+                            GError    **error)
+{
+  g_autoptr(GPtrArray) resolved_flatpaks = g_ptr_array_new_with_free_func (flatpak_to_install_free);
+  g_autofree gchar *explicitly_removed_flatpaks_list_path = g_build_filename (get_gnome_software_state_path (),
```

Comment:
We talked about this today and concluded that it would be best if this was provided by Flatpak (per-installation), to also cover the case when applications are removed or installed using another tool than gnome-software.

In general, we thought that the right approach for upstream is something along the lines of a transaction journal for Flatpak, and not make it only about removals. @smspillaz is going to talk to Alex and see what he thinks.

Reply:
That sounds like a very good idea.